src/github.com/mozilla-services/pushgo/simplepush/dynamo_store.go

```diff
@@ -0,0 +1,542 @@
+				vers = vera.Value
+			} else {
+				vers = ddb_getNow(0)
+			}
+		}
+		version, err := strconv.ParseUint(vers, 10, 64)
+		if err != nil {
+			version = uint64(time.Now().UTC().Unix())
+		}
+		updates = append(updates, Update{
+			ChannelID: r[CHID_LABEL].Value,
+			Version:   version,
+		})
+	}
+
+	return
+}
+
+// DropAll removes all channel records for the given device ID. Implements
+// Store.DropAll().
+func (s *DynamoDBStore) DropAll(uaid string) (err error) {
+	//TODO: Pass this to a goroutine to make it non blocking.
+	result, err := s.table.Query([]dynamodb.AttributeComparison{
+		*dynamodb.NewEqualStringAttributeComparison(UAID_LABEL, uaid),
+	})
+	if err != nil {
+		return
+	}
+	// Need to reflect to prevent testing from causing panics galore.
+	// Damn good thing this method is not on the critical path.
+	testing := reflect.TypeOf(s.table).String() == "*simplepush.testDynamoTable"
+	items := make(map[string][][]dynamodb.Attribute)
+	items["Delete"] = make([][]dynamodb.Attribute, len(result))
+	for n, record := range result {
+		//Use partial names for item labels (e.g. Put or Delete instead of
+		// "PutRequest" or "DeleteRequest")
+		items["Delete"][n] = []dynamodb.Attribute{
+			dynamodb.Attribute{
+				Type:  dynamodb.TYPE_STRING,
+				Name:  UAID_LABEL,
+				Value: record[UAID_LABEL].Value,
+			},
+			dynamodb.Attribute{
+				Type:  dynamodb.TYPE_STRING,
+				Name:  CHID_LABEL,
+				Value: record[CHID_LABEL].Value,
+			}}
+	}
+	defer func() {
+		// unravelling this is.. not simple or well explained.
+		// added bonus that there's no clear example of how to do it.
+		// So this will probably puke or throw exceptions like an angry
+		// monkey.
+		if r := recover(); r != nil {
+			if err, ok := r.(error); ok {
+				if s.logger.ShouldLog(ERROR) {
+					s.logger.Error("dynamodb", "Unhandled exception during dropall",
+						LogFields{"error": err.Error()})
+				}
+			}
+		}
+	}()
+	var i int
+	for i = 0; i < 5; i++ {
+		//AMZ states that it's possible for a BatchWrite to partially fail.
+		// In that case, you're required to retry the "missed" records after
+		// some period, ideally, with an exponential timeout
+		// I'm also arbitrarily giving up after 5 retries to prevent endless
+		// loops if crap has really gone to hell.
+		batch := s.table.BatchWriteItems(items)
+
+		missed := make(map[string]interface{})
+		if !testing {
+			missed, err = batch.Execute()
+		}
+		if err != nil {
+			if s.logger.ShouldLog(ERROR) {
+				s.logger.Error("dynamodb", "Could not complete DropAll",
+					LogFields{"error": err.Error()})
+			}
+			return err
+		}
+		if missActions, ok := missed["DeleteRequest"]; ok {
+
+			// This may throw an index errorfor various reasons.
+			items = missActions.(dynamodb.BatchWriteItem).ItemActions[s.table.(*dynamodb.Table)]
+			if len(items) > 0 {
+				// TODO: replace this with a proper exponential backoff.
+				time.Sleep(time.Duration(rand.Intn(4)+1) * time.Second)
+				continue
+			}
+		}
+		break
+	}
```

Comment:
Break this into a separate function that takes a table and items to batch write, so we can re-use it easily (in the dynamodb file).